File: app/src/main/java/com/android/medianet/launcher/LauncherDataSource.java
package com.android.medianet.launcher;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;


import java.util.ArrayList;
import java.util.List;

public class LauncherDataSource {
    private final ContentResolver resolver;

    public LauncherDataSource(Context context) {
        this.resolver = context.getContentResolver();
    }

    public List<ApplicationInfo> getAllApps() {
        List<ApplicationInfo> apps = new ArrayList<>();
        Cursor c = resolver.query(LauncherProvider.CONTENT_URI, null, null, null,
                "title COLLATE NOCASE ASC");
        if (c != null) {
            while (c.moveToNext()) {
                ApplicationInfo ai = cursorToApp(c);
                apps.add(ai);
            }
            c.close();
        }
        return apps;
    }

    public List<ApplicationInfo> getWorkspaceApps() {
        List<ApplicationInfo> apps = new ArrayList<>();
        Cursor c = resolver.query(LauncherProvider.CONTENT_URI, null, "container=?", new String[]{"workspace"},
                "screen ASC, cellY ASC, cellX ASC");
        if (c != null) {
            while (c.moveToNext()) apps.add(cursorToApp(c));
            c.close();
        }
        return apps;
    }

    public List<ApplicationInfo> getHotseatApps() {
        List<ApplicationInfo> apps = new ArrayList<>();
        Cursor c = resolver.query(LauncherProvider.CONTENT_URI, null, "container=?", new String[]{"hotseat"},
                "cellX ASC");
        if (c != null) {
            while (c.moveToNext()) apps.add(cursorToApp(c));
            c.close();
        }
        return apps;
    }

    public void insertApp(ApplicationInfo app) {
        ContentValues v = appToContentValues(app);
        resolver.insert(LauncherProvider.CONTENT_URI, v);
    }

    public void updateApp(ApplicationInfo app) {
        ContentValues v = new ContentValues();
        v.put("screen", app.screen);
        v.put("cellX", app.cellX);
        v.put("cellY", app.cellY);
        v.put("container", app.container);
        String sel = "_id=?";
        String[] args = { String.valueOf(app.id) };
        resolver.update(LauncherProvider.CONTENT_URI, v, sel, args);
    }

    public void deleteApp(ApplicationInfo app) {
        String sel = "_id=?";
        String[] args = { String.valueOf(app.id) };
        resolver.delete(LauncherProvider.CONTENT_URI, sel, args);
    }

    public boolean isFavoritesEmpty() {
        Cursor c = resolver.query(LauncherProvider.CONTENT_URI, new String[]{"_id"}, null, null, null);
        boolean empty = true;
        if (c != null) {
            empty = !c.moveToFirst();
            c.close();
        }
        return empty;
    }

    private ApplicationInfo cursorToApp(Cursor c) {
        ApplicationInfo ai = new ApplicationInfo();
        ai.id = c.getLong(c.getColumnIndex("_id"));
        ai.title = c.getString(c.getColumnIndex("title"));
        ai.screen = c.getInt(c.getColumnIndex("screen"));
        ai.cellX = c.getInt(c.getColumnIndex("cellX"));
        ai.cellY = c.getInt(c.getColumnIndex("cellY"));
        ai.container = c.getString(c.getColumnIndex("container"));
        // intent leaving null for now (we can parse if stored)
        return ai;
    }

    private ContentValues appToContentValues(ApplicationInfo app) {
        ContentValues v = new ContentValues();
        v.put("title", app.title);
        v.put("container", app.container);
        v.put("screen", app.screen);
        v.put("cellX", app.cellX);
        v.put("cellY", app.cellY);
        v.put("intent", app.intent != null ? app.intent.toUri(0) : "");
        return v;
    }
}

File: app/src/main/java/com/android/medianet/launcher/data/db/LauncherDataSource.kt
package com.android.medianet.launcher.data.db

import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.util.Log
import com.android.medianet.launcher.data.db.LauncherSettings.Favorites
import com.android.medianet.launcher.data.model.FolderInfo
import com.android.medianet.launcher.data.model.ItemInfo
import com.android.medianet.launcher.middlelayer.AllAppsList
import com.android.medianet.launcher.middlelayer.IconCache
import com.android.medianet.launcher.middlelayer.ShortcutInfo
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import java.net.URISyntaxException

class LauncherDataSource(val context: Context, val iconCache: IconCache, val allAppsList: AllAppsList) {


    fun loadWorkspace() : Flow<WorkspaceItems>{
        return flow{
            emit(loadWorkspaceItemsFromDatabase())
        }.flowOn(Dispatchers.IO)
    }


    private fun loadWorkspaceItemsFromDatabase() : WorkspaceItems{
        val listofItems = ArrayList<ItemInfo>()
        val folderMap = mutableMapOf<Long, FolderInfo>()
        val listofShortcuts = ArrayList<ShortcutInfo>()
        val itemsToRemove = ArrayList<Long>()

        val occupied = Array(2){
            Array(4){
                arrayOfNulls<ItemInfo>(4)
            }
        }

        val uri : Uri = Uri.parse("")

        context.contentResolver.query(uri, null, null,null,null)?.use {

            val itemTypeIndex = it.getColumnIndexOrThrow(LauncherSettings.Favorites.ITEM_TYPE)

            try {

                while(it.moveToNext()){
                    val itemType = it.getInt(itemTypeIndex)

                    when (itemType) {
                        Favorites.ITEM_TYPE_APPLICATION, Favorites.ITEM_TYPE_SHORTCUT -> retrieveShortcutData(
                            it,
                            listofShortcuts,
                            listofItems,
                            folderMap,
                            itemType,
                            occupied
                        )
                    }
                }
            }catch (e : Exception){
                Log.e(TAG, "loadWorkspaceItemsFromDatabase(): error occurred=${e.message}")
            }
        }

        val workspaceItems = WorkspaceItems(listofItems, folderMap, listofShortcuts)
        return workspaceItems
    }

    fun getShortcutInfo(intent: Intent, context: Context, c : Cursor) : ShortcutInfo{
        val info = ShortcutInfo()

        var icon : Bitmap? = null
        val titleIndex = c.getColumnIndexOrThrow(Favorites.TITLE)
        val pkgMgr = context.packageManager
        val resolveInfo = pkgMgr.resolveActivity(intent, 0)
        resolveInfo?.let {
            icon = iconCache.getIcon(intent.component , it)
        }

        if(icon == null) {
            icon = getIconFromCursor(c)
        }

        icon?.let {
            info.setIcon(it)
        }

        if(info.title == null){
            info.title = intent.component?.className

        }

        intent.component?.let {
            info.packagename = it.className
        }
        return info
    }

    fun getIconFromCursor(cursor: Cursor): Bitmap? {
        val iconIndex = cursor.getColumnIndexOrThrow(Favorites.ICON)
        val data = cursor.getBlob(iconIndex)
        return try {
            BitmapFactory.decodeByteArray(data, 0, data.size)
        } catch (e: Exception) {
            Log.e("", "")
            null
        }
    }



    private fun retrieveShortcutData(
        cursor: Cursor,
        listofShortcuts: ArrayList<ShortcutInfo>,
        listofItems : ArrayList<ItemInfo>,
        folderMap : MutableMap<Long,FolderInfo>,
        ItemType : Int,
        occupied : Array<Array<Array<ItemInfo?>>>){

        val intentIndex = cursor.getColumnIndexOrThrow(Favorites.INTENT)
        val idIndex = cursor.getColumnIndexOrThrow(Favorites._ID)
        val containerIndex = cursor.getColumnIndexOrThrow(Favorites.CONTAINER)
        val screenIndex = cursor.getColumnIndexOrThrow(Favorites.SCREEN)
        val cellXIndex = cursor.getColumnIndexOrThrow(Favorites.CELLX)
        val cellyIndex = cursor.getColumnIndexOrThrow(Favorites.CELLY)
        //val packageNameIndex = cursor.getColumnIndexOrThrow()
        val intentDescription = cursor.getString(intentIndex)

        var intent: Intent

        try {
            intent = Intent.parseUri(intentDescription, 0)
        } catch (e: URISyntaxException) {
            Log.e("", "$(Launcher Constants.PROFILE RETRIEVAL FLOW TAG) STAG Cannot parse the intent se")
            return
        }
        val shortcutInfo  = getShortcutInfo(intent , context, cursor)

        shortcutInfo.let {

            // updateSavedIcon(it, oursor)

            it.intent = intent
            it.id = cursor.getLong(idIndex)
            it.container = cursor.getInt(containerIndex)
            it.screenId = cursor.getInt(screenIndex)
            it.cellX = cursor.getInt(cellXIndex)
            it.cellY = cursor.getInt(cellyIndex)

            listofShortcuts.add(it)

            when(it.container){
                Favorites.CONTAINER_DESKTOP, Favorites.CONTAINER_HOTSEAT -> listofItems.add(it)
                else -> {
                    //folder section
                }
            }
        }
    }
}